client/scripts/controllers/chain/ethereum/marlin/adapter.ts

```diff
@@ -0,0 +1,111 @@
+import { MarlinTypes } from '@commonwealth/chain-events';
+import MarlinAPI from './api';
+import MarlinChain from './chain';
+import MarlinGovernance from './governance';
+import MarlinProposal from './proposal';
+import MarlinHolders from './holders';
+
+export default class Marlin extends IChainAdapter<EthereumCoin, EthereumAccount> {
+  public readonly base = ChainBase.Ethereum;
+  public readonly class = ChainClass.Marlin;
+  public chain: MarlinChain;
+  // public ethAccounts: EthereumAccounts;
+  // public accounts: MarlinHolders;
+  public accounts: EthereumAccounts;
+  public marlinAccounts:  MarlinHolders;
+  public governance: MarlinGovernance;
+  public readonly webWallet: EthWebWalletController = new EthWebWalletController();
+  public readonly chainEntities = new ChainEntityController();
+
+  constructor(meta: NodeInfo, app: IApp) {
+    super(meta, app);
+    this.chain = new MarlinChain(this.app);
+    this.accounts = new EthereumAccounts(this.app);
+    this.marlinAccounts = new MarlinHolders(this.app, this.chain, this.accounts);
+    this.governance = new MarlinGovernance(this.app);
+    this.accounts.init(this.chain);
+  }
+
+  public handleEntityUpdate(entity: ChainEntity, event: ChainEvent): void {
+    switch (entity.type) {
+      case MarlinTypes.EntityKind.Proposal: {
+        this.governance.updateProposal(entity, event);
+        break;
+      }
+      default: {
+        console.error('Received invalid marlin chain entity!');
+        break;
+      }
+    }
+  }
+
+  public async initApi() {
+    console.log('Marlin initApi()');
+    await this.chain.resetApi(this.meta);
+    await this.chain.initMetadata();
+    // await this.ethAccounts.init(this.chain);
+
+    const activeAddress: string = this.webWallet.accounts && this.webWallet.accounts[0];
+    const mpondContractAddress = this.meta.address;
+    const governorAlphaContractAddress = '0x777992c2E4EDF704e49680468a9299C6679e37F6';
+    const api = new MarlinAPI(this.meta.address, governorAlphaContractAddress, this.chain.api.currentProvider as any, activeAddress);
+    await api.init().catch((e) => {
+      this._failed = true;
+      notifyError('Please change your Metamask network');
+    });
+    this.chain.marlinApi = api;
+
+    if ((window as any).ethereum || (window as any).web3) {
+      await this.webWallet.enable();
+      await this.webWallet.enable().catch((e) => { console.log('thowing!22'); console.error(e); });
+      await this.webWallet.web3.givenProvider.on('accountsChanged', (accounts) => {
+        const updatedAddress = this.app.user.activeAccounts.find((addr) => addr.address === accounts[0]);
+        setActiveAccount(updatedAddress);
+      });
+
+      await this.webWallet.web3.givenProvider.on('accountsChanged', (accounts) => {
+        const updatedAddress = this.app.user.activeAccounts.find((addr) => addr.address === accounts[0]);
+        setActiveAccount(updatedAddress);
+        api.updateSigner(accounts[0]);
+      });
```

Comment:
Repeat? @zakhap

Reply:
The first 'accountsChanged' seems unnecessary, and missed the updateSigner call

Reply:
👍